src/sql/parts/query/editor/messagePanel.ts

```diff
@@ -145,27 +147,37 @@ export class MessagePanel extends ViewletPanel {
 		let hasError = false;
 		if (isArray(message)) {
 			hasError = message.find(e => e.isError) ? true : false;
+			message.forEach(resultMessage => this.countMessageLines(resultMessage));
 			this.model.messages.push(...message);
 		} else {
 			hasError = message.isError;
+			this.countMessageLines(message);
 			this.model.messages.push(message);
 		}
 		if (hasError) {
 			this.setExpanded(true);
 		}
 		if (this.state.scrollPosition) {
 			this.tree.refresh(this.model).then(() => {
-				this.tree.setScrollPosition(1);
+				// Restore the previous scroll position when switching between tabs
+				this.tree.setScrollPosition(this.state.scrollPosition);
 			});
 		} else {
 			const previousScrollPosition = this.tree.getScrollPosition();
 			this.tree.refresh(this.model).then(() => {
+				// Scroll to the end if the user was already at the end otherwise leave the current scroll position
 				if (previousScrollPosition === 1) {
 					this.tree.setScrollPosition(1);
 				}
 			});
 		}
-		this.maximumBodySize = this.model.messages.length * 22;
+	}
+
+	private countMessageLines(resultMessage: IResultMessage): void {
+		let lines = resultMessage.message.split('\n').length;
+		this.messageLineCountMap.set(resultMessage, lines);
+		this.totalLines += lines;
+		this.maximumBodySize = this.totalLines * 22;
```

Comment:
Does it make more sense to change the maximum body size after doing all the counts rather than after each?

Reply:
Good idea, just fixed it. Does everything else look good?